srv/src/main/java/my/bookshop/handlers/HierarchyHandler.java

```diff
@@ -0,0 +1,236 @@
+            return topLevels;
+        }
+        return null;
+    }
+
+    private void setResult(CdsReadEventContext event, List<GenreHierarchy> result) {
+        if (!result.isEmpty()) {
+            addDrillState(result);
+        }
+
+        event.setResult(result);
+    }
+
+    private void addDrillState(List<GenreHierarchy> ghs) {
+        List<Integer> ids = ghs.stream().map(gh -> gh.getNodeId()).toList();
+        Set<Integer> parents = ghs.stream().map(gh -> gh.getParentId()).filter(p -> p != 0).collect(Collectors.toSet());
+        CqnSelect q = Select.from(GENRE_HIERARCHY, gh -> gh.parent()).columns(gh -> gh.node_id())
+                .where(gh -> gh.node_id().in(ids));
+        Set<Object> nonLeafs = db
+                .run(q)
+                .stream().map(r -> r.get(GenreHierarchy.NODE_ID)).collect(Collectors.toSet());
+
+        for (GenreHierarchy gh : ghs) {
+            Integer id = gh.getNodeId();
+            if (nonLeafs.contains(id)) {
+                if (parents.contains(id)) {
+                    gh.setDrillState("expanded");
+                } else {
+                    gh.setDrillState("collapsed");
+                }
+            } else {
+                gh.setDrillState("leaf");
+            }
+        } 
+    }
+        
+    private List<GenreHierarchy> handleDescendants(CqnDescendantsTransformation descendants) {
+        CqnFilterTransformation filter = (CqnFilterTransformation) descendants.transformations().get(0);
+
+        CqnPredicate parentFilter = CQL.copy(filter.filter(), new Modifier() {
+            @Override
+            public CqnValue ref(CqnElementRef ref) {
+                return CQL.get(GenreHierarchy.PARENT_ID);
+            }
+        });
+
+        CqnSelect childrenCQN = Select.from(GENRE_HIERARCHY).where(parentFilter);
+        List<GenreHierarchy> children = db.run(childrenCQN).listOf(GenreHierarchy.class);
+
+        return children.stream().sorted(new Sorter()).toList();
+    }
+
+    private List<GenreHierarchy> handleAncestors(CqnAncestorsTransformation ancestors, CqnTopLevelsTransformation topLevels) {
+        CqnTransformation trafo = ancestors.transformations().get(0);
+        Select<GenreHierarchy_> inner = Select.from(GENRE_HIERARCHY).where(gh -> gh.node_id().eq(CQL.get("$outer.node_id")));
+        if (trafo instanceof CqnFilterTransformation filter) {
+            inner.where(CQL.and(filter.filter(), inner.where().get()));
+        } else if (trafo instanceof CqnSearchTransformation search) {
+            inner.search(search.search());
+        }
+        Select<GenreHierarchy_> outer = Select.from(GENRE_HIERARCHY).columns(gh -> gh.node_id().as("i0"), gh -> gh.parent().node_id().as("i1"),
+            gh -> gh.parent().parent().node_id().as("i2"), gh -> gh.parent().parent().parent().node_id().as("i3"),
+            gh -> gh.parent().parent().parent().parent().node_id().as("i4")).where(gh -> gh.exists($outer -> inner));
```

Comment:
or using an in-subquery:

```suggestion
        Select<GenreHierarchy_> inner = Select.from(GENRE_HIERARCHY).columns(gh -> gh.node_id());
        if (trafo instanceof CqnFilterTransformation filter) {
            inner.where(filter.filter());
        } else if (trafo instanceof CqnSearchTransformation search) {
            inner.search(search.search());
        }
        Select<GenreHierarchy_> outer = Select.from(GENRE_HIERARCHY).columns(gh -> gh.node_id().as("i0"), gh -> gh.parent().node_id().as("i1"),
            gh -> gh.parent().parent().node_id().as("i2"), gh -> gh.parent().parent().parent().node_id().as("i3"),
            gh -> gh.parent().parent().parent().parent().node_id().as("i4")).where(gh -> gh.node_id().in(inner));
```